integration_test/tests/disconnect_tests.rs

```diff
@@ -0,0 +1,130 @@
+        let query = b"SELECT * FROM integration_test.foo\0".map(|b| b as u16);
+        unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(stmt, query.as_ptr(), SQL_NTS as i32,),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            // we didn't actually bind a statement, so this should be a no-op
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt, FreeStmtOption::SQL_UNBIND as i16),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+
+            // we will not free the statement, instead proceeding directly to disconnect and freeing the dbc and env handles
+            disconnect_and_free_dbc_and_env_handles(env, dbc)
+        }
+    }
+
+    #[test]
+    fn sql_disconnect_handles_many_statements_properly() {
+        let env = allocate_env(AttrOdbcVersion::SQL_OV_ODBC3);
+        let dbc = connect_with_conn_string(env, None)
+            .expect("Failed to connect with default connection string");
+        let stmt_1 = allocate_statement(dbc).expect("Failed to allocate statement 1");
+        let stmt_2 = allocate_statement(dbc).expect("Failed to allocate statement 2");
+        let stmt_3 = allocate_statement(dbc).expect("Failed to allocate statement 3");
+        let stmt_4 = allocate_statement(dbc).expect("Failed to allocate statement 4");
+        let stmt_5 = allocate_statement(dbc).expect("Failed to allocate statement 5");
+        let statements = [stmt_1, stmt_2, stmt_3, stmt_4, stmt_5];
+
+        let query = b"SELECT * FROM integration_test.foo\0".map(|b| b as u16);
+        statements.iter().for_each(|stmt| unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLExecDirectW(*stmt, query.as_ptr(), SQL_NTS as i32,),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt.cast::<Handle>()),
+            );
+        });
+
+        // we will close statements 1 and 3
+        unsafe {
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_1, FreeStmtOption::SQL_CLOSE as i16),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt_1.cast::<Handle>()),
+            );
+
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                SQLFreeStmt(stmt_3, FreeStmtOption::SQL_CLOSE as i16),
+                "{}",
+                get_sql_diagnostics(HandleType::SQL_HANDLE_STMT, *stmt_3.cast::<Handle>()),
+            );
+
+            // and we'll free statement 3 handle
+            assert_eq!(
+                SqlReturn::SUCCESS,
+                // interestingly, rust doesn't like the cast here, so we'll just use the raw value
```

Comment:
```suggestion
```
It's unclear what this comment is referring to -- do you mean it doesn't like some cast that isn't included in the code? If so, what would that cast be and why is it important? If it is rejected by the compiler, doesn't that mean it's likely invalid? I think we may want to remove this comment before merging.

Reply:
It's referring to the use of `as *mut _` instead of `.cast::<Handle>`

Reply:
Actually I'm glad you flagged this. In further testing, it's actually unsafe to cast on the pointer. `as *mut _` apparently takes the address of the pointer, which is what is needed, otherwise there's an access violation. `*ptr::addr_of(x)` does the trick nicely and is "safer" than `as`.